activesupport/lib/active_support/json_with_marshal_fallback.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  class JsonWithMarshalFallback
+    MARSHAL_SIGNATURE = "\x04\x08"
+
+    cattr_accessor :fallback_to_marshal_deserialization, instance_accessor: false, default: true
+    cattr_accessor :use_marshal_serialization, instance_accessor: false, default: true
+
+    class << self
+      def dump(value)
+        if self.use_marshal_serialization
+          Marshal.dump(value)
+        else
+          JSON.encode(value)
+        end
+      end
+
+      def load(value)
+        raise ::JSON::ParserError if value.start_with?(MARSHAL_SIGNATURE)
+        JSON.decode(value)
+      rescue ::JSON::ParserError
+        if self.fallback_to_marshal_deserialization
+          Marshal.load(value)
```

Comment:
Should the fallback to Marshal be logged?

Reply:
Ya, I think that makes sense. Hopefully it goes to 0, but it's probably an interesting metric for folks to capture.